### PR TITLE
NEW: Signal.unfolded context manager

### PR DESCRIPTION
**Pull request by vidartf**
Suggested implementation of #615, resubmitted from #621.

### TODO
 - [x] Tests for new keywords to `Signal.unfold()` / `Signal.unfolded()`
 - [x] Use in appropriate places in existing code
 - [x] Address questions marked with TODOs

### CODE REVIEW

File: hyperspy/learn/mva.py [hunk 1]
```diff
@@ -814,15 +814,17 @@ def _calculate_recmatrix(self, components=None, mva_type=None,):
...
+            sc = self.deepcopy()
+            sc.data = a.T.reshape(self.data.shape)
+            sc.metadata.General.title += signal_name
+            if target.mean is not None:
+                sc.data += target.mean
+        finally:
+            if self._unfolded4decomposition is True:
+                self.fold()
+                sc.fold()
+            # TODO: Should self._unfolded4decomposition be set to False?
```
Comment (vidartf) [inline hyperspy/learn/mva.py]: The old code never set `_unfolded4decomposition` to `False` here, should it be done?

File: hyperspy/misc/utils.py [hunk 2]
```diff
@@ -101,12 +102,10 @@ def unfold_if_multidim(signal):
...
-    if len(signal.axes_manager._axes) > 2:
-        print "Automatically unfolding the SI"
-        signal.unfold()
-        return True
-    else:
-        return False
+    import warnings
+    warnings.warn("unfold_if_multidim is deprecated and will be removed in "
+                  "0.9 please use Signal.unfold instead")
+    return None
```
Comment (vidartf) [inline hyperspy/misc/utils.py]: Also deprecated this function. I'll PR it separate if the need arises.

File: hyperspy/learn/mva.py (hunk 1 above)
Comment (francisco-dlp) [inline hyperspy/learn/mva.py]: Although it makes not difference now, it's probably good practice to set it back to False.

File: hyperspy/misc/utils.py [hunk 3]
```diff
@@ -101,12 +102,10 @@ def unfold_if_multidim(signal):
...
     """
-    if len(signal.axes_manager._axes) > 2:
-        print "Automatically unfolding the SI"
-        signal.unfold()
-        return True
-    else:
-        return False
+    import warnings
+    warnings.warn("unfold_if_multidim is deprecated and will be removed in "
+                  "0.9 please use Signal.unfold instead")
```
Comment (francisco-dlp) [inline hyperspy/misc/utils.py]: This should probably be:
```python
    warnings.warn("unfold_if_multidim is deprecated and will be removed in "
                  "0.9 please use Signal.unfold instead", DeprecationWarning)

```

**Comment by to266** [comment]
@vidartf , do you think you'll be able to finish this PR until tomorrows release? If not, feel free to change the milestone

**Comment by vidartf** [comment]
Should be OK for review.

**Pull request merged**